### PR TITLE
SDK - Containers - Do not create GCS bucket unless building the image

### PR DESCRIPTION
**Pull request by Ark-kun**
Previously, when the user tried to construct `ContainerBuilder` it immediately created a storage bucket even is no image was build. This prevented the creation of default builder instance.

Also missing default image name is no longer an error as long as the image name is provided to ContainerBuilder.build.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/1938)

### CODE REVIEW

**Comment by Ark-kun** [comment]
@numerology I've talked to @gaoning777 and he agreed with the idea of this refactoring.

File: sdk/python/kfp/compiler/_container_builder.py [hunk 1]
```diff
@@ -22,86 +22,91 @@
...
   ContainerBuilder helps build a container image
   """
-  def __init__(self, gcs_staging=None, gcr_image_tag=None, namespace=None):
+  def __init__(self, gcs_staging=None, default_image_name=None, namespace=None):
     """
     Args:
       gcs_staging (str): GCS bucket/blob that can store temporary build files,
           default is gs://PROJECT_ID/kfp_container_build_staging.
-      gcr_image_tag (str): GCR image tag where the target image is pushed
+      default_image_name (str): GCR image tag where the target image is pushed
```
Comment (numerology) [inline sdk/python/kfp/compiler/_container_builder.py]: `default_image_name (str): GCR image tag where the target image is pushed` Does the description need to update, too?

**Comment by numerology** [comment]
/lgtm

Left one nit comment.

**Comment by gaoning777** [comment]
> @numerology I've talked to @gaoning777 and he agreed with the idea of this refactoring.

Just for the record, I'm not strongly opinionated about whether fetching the project id/creating bucket in the constructor or in the build function. 
However, this refactoring does change the constructor signature(default_image_name). I understand that this class is not exposed in the __init__.py but there might still be users using it. I'm suggesting to not break that change.

File: sdk/python/kfp/compiler/_container_builder.py (hunk 1 above)
Comment (Ark-kun) [reply sdk/python/kfp/compiler/_container_builder.py]: Thanks for the catch. Fixed.

**Comment by Ark-kun** [comment]
> However, this refactoring does change the constructor signature(default_image_name). I understand that this class is not exposed in the **init**.py but there might still be users using it. I'm suggesting to not break that change.

Sure. I'll split that change out.

**Comment by Ark-kun** [comment]
> However, this refactoring does change the constructor signature(default_image_name).

Changed that back. @gaoning777

**Comment by gaoning777** [comment]
Do you want to merge with master? Looks like the samples tests are failing consistently.

**Comment by Ark-kun** [comment]
> Do you want to merge with master? Looks like the samples tests are failing consistently.

I've checked and these were temporary issues with PyPI network.

**Pull request merged**